### PR TITLE
Make GenericRow support null value fields and default null values

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
This is part of the null value support.
In order to support null values, we can store the fields with null values in the GenericRow, and fill the defaultNullValue to the fields.
The reason why null value fields are not stored as a special field is that the caller might mix the special field with regular field, and storing a separate set can avoid the extra special field handling.

- Add a set in GenericRow to store the null value fields
- Redesign the APIs for GenericRow for a cleaner contract
- Modify NullValueTransformer to fill default null values to null fields
- Fix the RecordTransformer to not transform virtual column
- Enhance the RecordTransformerTest

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 1]
```diff
@@ -18,85 +18,148 @@
...
+ * The generic row is the value holder returned from {@link RecordReader#next()} and
+ * {@link RecordReader#next(GenericRow)}, and can be modified with {@link RecordTransformer}. The generic row returned
+ * from the {@link NullValueTransformer} should have {@code defaultNullValue} filled to the fields with {@code null}
+ * value, so that for fields with {@code null} value, {@link #getValue(String)} will return the {@code defaultNullValue}
+ * and {@link #isNullValue(String)} will return {@code true}.
  */
-public class GenericRow implements RowEvent {
-  private Map<String, Object> _fieldMap = new HashMap<>();
+public class GenericRow {
+  private final Map<String, Object> _fieldToValueMap = new Object2ObjectOpenHashMap<>();
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: is fastutil really needed? at some point we want to move GenericRow and RecordReader out of pinot core and part of connectors (batch and real-time) into a separate module that external users can use. Its better to keep the dependency to a minimum.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: Make sense. Fastutil map is faster, but maybe not necessary. Will replace with java map.

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 2]
```diff
@@ -18,85 +18,148 @@
...
-  @Override
-  public String[] getFieldNames() {
-    return _fieldMap.keySet().toArray(new String[_fieldMap.size()]);
+  /**
+   * Returns the fields with {@code null} value.
+   * <p>The {@code nullField} will be set when setting the {@code nullDefaultValue} for field by calling
+   * {@link #putDefaultNullValue(String, Object)}.
+   */
+  public Set<String> getNullValueFields() {
+    return _nullValueFields;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 3]
```diff
@@ -18,85 +18,148 @@
...
 
-  public Set<Map.Entry<String, Object>> getEntrySet() {
-    return _fieldMap.entrySet();
+  /**
+   * Returns the map from fields to values.
+   * <p>Before setting the {@code defaultNullValue} for a field by calling {@link #putDefaultNullValue(String, Object)},
+   * the value for the field can be {@code null}.
+   */
+  public Map<String, Object> getFieldToValueMap() {
+    return _fieldToValueMap;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 4]
```diff
@@ -18,85 +18,148 @@
...
-      } else {
-        b.append(value);
-      }
-      b.append(", ");
-    }
-    return b.toString();
+  /**
+   * Sets the value for the given field.
+   */
+  public void putValue(String fieldName, @Nullable Object value) {
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: value is nullable. sometimes we might put null value, and NullValueTransformer will replace it with null default value.

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 5]
```diff
@@ -18,85 +18,148 @@
...
+ * {@link RecordReader#next(GenericRow)}, and can be modified with {@link RecordTransformer}. The generic row returned
+ * from the {@link NullValueTransformer} should have {@code defaultNullValue} filled to the fields with {@code null}
+ * value, so that for fields with {@code null} value, {@link #getValue(String)} will return the {@code defaultNullValue}
+ * and {@link #isNullValue(String)} will return {@code true}.
  */
-public class GenericRow implements RowEvent {
-  private Map<String, Object> _fieldMap = new HashMap<>();
+public class GenericRow {
+  private final Map<String, Object> _fieldToValueMap = new HashMap<>();
+  private final Set<String> _nullValueFields = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: I prefer `_nullValueFields` haha

File: pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java [hunk 6]
```diff
@@ -18,85 +18,148 @@
...
 
-  @Override
-  public void init(Map<String, Object> field) {
-    _fieldMap = field;
+  /**
+   * Initializes the generic row from the given generic row (shallow copy).
+   */
+  public void init(GenericRow row) {
+    _fieldToValueMap.putAll(row._fieldToValueMap);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/GenericRow.java]: The row should be new created or cleared before calling init. Added javadoc for this. Do not explicit clear for performance concern.

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java [hunk 7]
```diff
@@ -58,60 +58,49 @@
...
 
-  private final Schema _schema;
   private final Map<String, PinotDataType> _dataTypes = new HashMap<>();
 
   public DataTypeTransformer(Schema schema) {
-    _schema = schema;
-    for (Map.Entry<String, FieldSpec> entry : schema.getFieldSpecMap().entrySet()) {
-      _dataTypes.put(entry.getKey(), PinotDataType.getPinotDataType(entry.getValue()));
+    for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
+      if (!fieldSpec.isVirtualColumn()) {
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java]: `schema.getPhysicalColumnNames()` is also looping over all the fieldSpecs, and create an extra set.
Directly looping over all the fieldSpecs gives better performance and produces less garbage.

**Pull request merged**